modules/migrate.py

```diff
@@ -0,0 +1,485 @@
+    """Returns a list of all HTML files under the root directory.
+
+    Args:
+        rootDir - String absolute path of the directory tree to search.
+    """
+    htmlFiles = []
+    for root, dirs, files in os.walk(rootDir):
+        for filename in [f for f in files if f.endswith('.html')]:
+            htmlFiles.append(os.path.join(root, filename))
+    return htmlFiles
+
+def _updateHTMLFile(filePath, widgetAbsolutePath, modularWidgetAbsolutePath):
+    """Updates file and linked widget JSON config files.
+
+    Updates the specified HTML file changing all references to the
+    non-modular widget to instead reference the modular widget, and all paths
+    relative to the non-modular widget to be relative to the modular widget.
+    Also fixes any JSON configuration files references from a widget's <object>
+    tag.
+    """
+    log.debug('Reading HTML file: %s', filePath)
+    with open(filePath, 'rb') as htmlFile:
+        lines = ''.join(htmlFile.readlines())
+        soup = BeautifulSoup(lines)
+
+        # Non-modular Widget view page path relative to the content. This path
+        # will be the data attribute of any widget object tags.
+        widgetViewRelativePath = os.path.join(
+            os.path.relpath(widgetAbsolutePath, os.path.dirname(filePath)),
+            'index.html')
+
+        # Absolute path to the modular version of the widget view page.
+        absModuleViewPath = os.path.join(
+            modularWidgetAbsolutePath, 'index.html')
+        # Widget view page path relative to the content.
+        relativeModuleViewPath = os.path.relpath(
+            absModuleViewPath, os.path.dirname(filePath))
+
+        # Convert all object data (iframe src) links to new modular location.
+        # This migrates all the widgets to point to their modular versions, but
+        # leaves all parameters and JSON config files incorrectly relative to
+        # the wrong location.
+        lines = re.sub(r'(<object\s.*?data=")' + widgetViewRelativePath +'"',
+                       r'\g<1>' + relativeModuleViewPath + '"',
+                       lines,
+                       flags=re.IGNORECASE)
+
+        # For all object tags, clean up params and including config file. Use
+        # the old relative path because BeautifulSoup parse tree was created
+        # before the regex replace above and doesn't know about it.
+        for object in soup.find_all('object',
+                attrs={'data': widgetViewRelativePath}):
+            # Find all param tags for that widget
+            for param in object.find_all('param'):
+                # Param value could be anything, but it might be a relative
+                # path. Treat it as such creating an absolute path and see if it
+                # exists. If it does, assume it is a relative path to something
+                # in the project and update it to be relative to the new modular
+                # widget location.
+                val = param['value']
```

Comment:
Param tags are always guaranteed to have value attributes right? I wouldn't be opposed of using `param.get('value')` instead in case someone has a really odd param tag without a value.